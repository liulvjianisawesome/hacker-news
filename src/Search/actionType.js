export const SET_SEARCHTERM = 'SEARCHTERM/SET';
export const SET_SEARCHKEY = 'SEARCHKEY/SET';
export const SET_ISLOADING = 'ISLOADING/SET';
export const SET_RESULTS = 'RESULTS/SET';
export const SET_ERROR = 'ERROR/SET';