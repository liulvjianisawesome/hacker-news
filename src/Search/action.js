import { SET_SEARCHTERM, SET_SEARCHKEY, SET_ISLOADING, SET_RESULTS, SET_ERROR } from './actionType';
import { DEFAULT_QUERY, DEFAULT_HPP, PATH_BATH, PATH_SEARCH, PARAM_SEARCH, PARAM_PAGE, param_HPP } from '../constant';

export const setSearchTerm = searchTerm => ({
  type: SET_SEARCHTERM,
  searchTerm: searchTerm,
});

export const setSearchKey = searchKey => ({
  type: SET_SEARCHKEY,
  searchKey: searchKey,
});

export const setIsLoading = isLoading => ({
  type: SET_ISLOADING,
  isLoading: isLoading,
});

export const setResults = results => ({
  type: SET_RESULTS,
  results: results,
});

export const setError = error => ({
  type: SET_ERROR,
  error: error,
});

export function fetchData(searchTerm = '', page = 0) {
  return (dispatch, getState) => {
    dispatch(setIsLoading(true));

    return fetch(`${PATH_BATH}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${param_HPP}${DEFAULT_HPP}`)
      .then(response => response.json())
      .then(result => {
        const { hits, page } = result;
        const { searchKey, results } = getState();

        const oldHits = (results && results[searchKey])
          ? results[searchKey].hits
          : [];

        const updatedHits = [
          ...oldHits,
          ...hits
        ];

        const lastResults = {
          ...results,
          [searchKey]: { hits: updatedHits, page },
        };
        dispatch(setResults(lastResults));

        dispatch(setIsLoading(false));
      })
      .catch(err => {
        dispatch(setError(err));
      })
  }
}