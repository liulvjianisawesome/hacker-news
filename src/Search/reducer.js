import { SET_SEARCHTERM, SET_SEARCHKEY, SET_ISLOADING, SET_RESULTS, SET_ERROR } from './actionType';
import { DEFAULT_QUERY } from '../constant';

export const searchTermReducer = (state=DEFAULT_QUERY, action) => {
  if (action.type === SET_SEARCHTERM) {
    return action.searchTerm;
  }
  return state;
};

export const searchKeyReducer = (state='', action) => {
  if (action.type === SET_SEARCHKEY) {
    return action.searchKey;
  }
  return state;
};

export const isLoadingReducer = (state=false, action) => {
  if (action.type === SET_ISLOADING) {
    return action.isLoading;
  }
  return state;
};

export const resultsReducer = (state=null, action) => {
  if (action.type === SET_RESULTS) {
    return action.results;  
  }
  return state;
};

export const errorReducer = (state=null, action) => {
  if (action.type === SET_ERROR) {
    return action.error;
  }
  return state;
}
