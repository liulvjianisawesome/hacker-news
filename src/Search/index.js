import React from 'react';
import PropTypes from 'prop-types';

import { setSearchTerm, setSearchKey, fetchData } from './action';
import { connect } from 'react-redux';

class Search extends React.Component {
  constructor(props) {
    super(props);

    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  onSearchChange(event) {
    this.props.dispatch(setSearchTerm(event.target.value));
  }

  onSubmit(event) {
    const { searchTerm, results } = this.props;
    this.props.dispatch(setSearchKey(searchTerm));

    if (!results[searchTerm]) {
      this.props.dispatch(fetchData(searchTerm))
    }

    event.preventDefault();
  }

  componentDidMount() {
    const { searchTerm } = this.props;
    this.props.dispatch(setSearchKey(searchTerm));
    this.props.dispatch(fetchData(searchTerm));
  }

  render() {
    return (
      <form onSubmit={this.onSubmit}>
        <button type="submit">
          {this.props.children}
        </button>
        <input type="text" onChange={this.onSearchChange} value={this.props.searchTerm} />
      </form>
    );
  }
}

Search.propTypes = {
  children: PropTypes.node,
};

const mapStateToProps = (state) => {
  return {
    searchTerm: state.searchTerm,
    results: state.results,
  };
}

export default connect(mapStateToProps, null)(Search);