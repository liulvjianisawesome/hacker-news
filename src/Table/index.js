import React from 'react';
import PropTypes from 'prop-types';
import { sortBy } from 'lodash';
import classNames from 'classnames';
import { connect } from 'react-redux';
import { setResults } from '../Search/action';

const SORTS = {
  NONE: list => list,
  TITLE: list => sortBy(list, 'title'),
  AUTHOR: list => sortBy(list, 'author'),
  COMMENTS: list => sortBy(list, 'num_comments').reverse(),
  POINTS: list => sortBy(list, 'points').reverse(),
}

const Sort = ({ sortKey, onSort, children, activeSortKey }) => {
  const sortClass = classNames(
    'button-inline',
    { 'button-active': sortKey === activeSortKey }
  )

  return (
    <button
      className={sortClass}
      onClick={() => onSort(sortKey)}>
      {children}
    </button>
  )
}

class Table extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      sortKey: 'NONE',
      isSortReverse: false,
    };

    this.onSort = this.onSort.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
  }

  onSort(sortKey) {
    const isSortReverse = sortKey === this.state.sortKey && !this.state.isSortReverse;

    this.setState({
      sortKey,
      isSortReverse,
    });
  }

  onDismiss(id) {
    const { results, searchKey } = this.props;
    const { hits, page } = results[searchKey];

    const updatedHits = hits.filter(item => item.objectID !== id);

    const lastResults = {
      ...results,
      [searchKey]: { hits: updatedHits, page },
    };

    this.props.dispatch(setResults(lastResults));
  }

  render() {
    const { list } = this.props;
    const { sortKey, isSortReverse } = this.state;
    const onSort = this.onSort;

    const finalList = isSortReverse ? SORTS[sortKey](list).reverse() : SORTS[sortKey](list);

    return (
      <div className="table">
        <div className="table-header">
          <span style={{ width: '40%' }}>
            <Sort
              sortKey={'TITLE'}
              onSort={onSort}
              activeSortKey={sortKey}
            >
              Title
            </Sort>
          </span>
          <span style={{ width: '30%' }}>
            <Sort
              sortKey={'AUTHOR'}
              onSort={onSort}
              activeSortKey={sortKey}
            >
              Author
            </Sort>
          </span>
          <span style={{ width: '10%' }}>
            <Sort
              sortKey={'COMMENTS'}
              onSort={onSort}
              activeSortKey={sortKey}
            >
              Comments
            </Sort>
          </span>
          <span style={{ width: '10%' }}>
            <Sort
              sortKey={'POINTS'}
              onSort={onSort}
              activeSortKey={sortKey}
            >
              Points
            </Sort>
          </span>
          <span style={{ width: '10%' }}>
          </span>
        </div>
        {
          finalList
            .map((item) => {
              return (
                <div key={item.objectID} className="table-row">
                  <span style={{ width: '40%' }}>
                    <a href={item.url}>{item.title}</a>
                  </span>
                  <span style={{ width: '30%' }}>{item.author}</span>
                  <span style={{ width: '10%' }}>{item.num_comments}</span>
                  <span style={{ width: '10%' }}>{item.points}</span>
                  <span style={{ width: '10%' }}>
                    <button className="button-inline" onClick={() => this.onDismiss(item.objectID)}>Dismiss</button>
                  </span>
                </div>
              );
            })
        }
      </div>
    )
  }
}

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      author: PropTypes.string,
      url: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number,
    })
  ).isRequired,
}

const mapStateToProps = (state) => {
  return {
    results: state.results,
    searchKey: state.searchKey,
    list: (state.results && state.results[state.searchKey] && state.results[state.searchKey].hits) || [],
  }
}

export default connect(mapStateToProps, null)(Table);