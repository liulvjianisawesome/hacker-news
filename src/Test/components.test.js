import React from 'react';
import ReactDOM from 'react-dom';
import 'isomorphic-fetch';

import { Provider } from 'react-redux';
import { createStore, combineReducers, applyMiddleware, compose } from 'redux';
import { searchTermReducer, searchKeyReducer, resultsReducer } from '../Search/reducer'
import thunk from 'redux-thunk';

import Enzyme, { shallow, mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import renderer from 'react-test-renderer';

import Search from '../Search';
import Table from '../Table';

Enzyme.configure({ adapter: new Adapter() });

describe('Search', () => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
  const store = createStoreWithMiddleware(combineReducers({
    searchTerm: searchTermReducer,
    results: resultsReducer,
  }));

  const subject = (
    <Provider store={store}>
      <Search>Search</Search>
    </Provider>
  );

  it('renders without crashing', () => {
    ReactDOM.render(subject, document.createElement('div'));
  });

  it('has a valid snapshot', () => {
    const component = renderer.create(subject);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Table', () => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);

  const list = [
    {
      title: '1',
      author: '1',
      num_comments: 1,
      points: 2,
      objectID: 'x',
    },
    {
      title: '2',
      author: '2',
      num_comments: 1,
      points: 2,
      objectID: 'y',
    },
  ]
  const searchKey = 'test';
  const results = {
    [searchKey]: { hits: list },
  };

  const store = createStoreWithMiddleware(combineReducers({
    searchKey: searchKeyReducer,
    results: resultsReducer,
  }), {
    searchKey: searchKey,
    results: results,
  });

  const subject = (
    <Provider store={store}>
      <Table />
    </Provider>
  );

  it('renders without crashing', () => {
    ReactDOM.render(subject, document.createElement('div'));
  });

  it('has a valid snapshot', () => {
    const component = renderer.create(subject);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows tow items in list', () => {
    const element = mount(subject);

    expect(element.find('.table-row').length).toBe(2);
  })
})