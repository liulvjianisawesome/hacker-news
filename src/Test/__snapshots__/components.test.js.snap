// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search has a valid snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <button
    type="submit"
  >
    Search
  </button>
  <input
    onChange={[Function]}
    type="text"
    value="redux"
  />
</form>
`;

exports[`Table has a valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
      >
        Title
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
      >
        Author
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
      >
        Comments
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
      >
        Points
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    />
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href={undefined}
      >
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        href={undefined}
      >
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
